### PR TITLE
Extract capturing before execution state

### PR DESCRIPTION
**Pull request by lptr**
With this change execution state before execution is now captured by the execution engine in a uniform way.

This change removes a major blocker from moving more stuff to the shared execution engine, enabling the moving of validation, snapshotting outputs and more.

### CODE REVIEW

**Comment by lptr** [comment]
@wolfs please take a look, I've rebased the branch on top of yours. Please ignore the first three commits, those will come out as soon as your branch is merged I think.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                        : fingerprintOutputSnapshot(outputSnapshot, previousOutputFingerprint, hasOverlappingOutputs);
+                }
+            )
+        );
+    }
+
+    private static CurrentFileCollectionFingerprint fingerprintOutputSnapshot(FileSystemSnapshot beforeExecutionOutputSnapshots, FileCollectionFingerprint previousOutputFingerprint, boolean hasOverlappingOutputs) {
+        List<FileSystemSnapshot> roots = hasOverlappingOutputs
+            ? OutputFilterUtil.filterOutputSnapshotBeforeExecution(previousOutputFingerprint, beforeExecutionOutputSnapshots)
+            : ImmutableList.of(beforeExecutionOutputSnapshots);
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java]: We probably should use a `CompositeFileSystemSnapshot` here as well instead of passing around a list.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy [hunk 2]
```diff
@@ -198,7 +247,7 @@ class ExecuteActionsTaskExecuterTest extends Specification {
         then:
         1 * standardOutputCapture.stop()
         then:
-        noMoreInteractions()
+        //noMoreInteractions()
```
Comment (wolfs) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy]: I think this should be enabled again, right?

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy [hunk 3]
```diff
@@ -66,39 +80,73 @@ import static org.gradle.internal.work.AsyncWorkTracker.ProjectLockRetention.REL
 
 class ExecuteActionsTaskExecuterTest extends Specification {
```
Comment (wolfs) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy]: This test is getting pretty strange now. Maybe we should replace it with a test which checks that the "correct" unit of work is generated without having the unit of work go completely through the execution engine. WDYT?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java [hunk 4]
```diff
@@ -42,13 +45,35 @@
...
+
     void visitOutputProperties(OutputPropertyVisitor visitor);
 
+    interface OutputPropertyVisitor {
+        void visitOutputProperty(String propertyName, TreeType type, FileCollection roots);
+    }
+
     void visitLocalState(LocalStateVisitor visitor);
 
-    @FunctionalInterface
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java]: Why did you remove the `@FunctionalInterface` annotations?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        Step<? super BeforeExecutionContext, ? extends CachingResult> delegate
+    ) {
+        this.classLoaderHierarchyHasher = classLoaderHierarchyHasher;
+        this.valueSnapshotter = valueSnapshotter;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public CachingResult execute(AfterPreviousExecutionContext context) {
+        BeforeExecutionState beforeExecutionState = context.getWork().isTaskHistoryMaintained()
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java]: Why don't we do:
```
        Optional<BeforeExecutionState> beforeExecutionState = context.getWork().isTaskHistoryMaintained()
            ? Optional.of(createExecutionState(context))
            : Optional.empty();
```

**Comment by lptr** [comment]
FTR, I think the three commits showed up because my PR was based on a later version of `master` than yours when I rebased it on top of it.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java [hunk 6]
```diff
@@ -234,6 +247,45 @@ public ExecutionHistoryStore getExecutionHistoryStore() {
...
+            List<InputChangesAwareTaskAction> taskActions = task.getTaskActions();
+            for (InputChangesAwareTaskAction taskAction : taskActions) {
+                visitor.visitAdditionalImplementation(taskAction.getActionImplementation(classLoaderHierarchyHasher));
+            }
+        }
+
+        @Override
+        public void visitInputProperties(InputPropertyVisitor visitor) {
+            Map<String, Object> inputPropertyValues = context.getTaskProperties().getInputPropertyValues().create();
+            assert inputPropertyValues != null;
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: Good point, done. (Though I hope that all of `TaskProperties` will be gone some time in the near future, along with this complexity.)

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java (hunk 1 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java]: Yes, though I'd like to address this in the next PR where we extract this (currently duplicated) code to another shared execution step.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy (hunk 2 above)
Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy]: Oops, fixed.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy (hunk 3 above)
Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy]: We do have a similar test that is building up from the other side: `IncrementalExecutionIntegrationTest`. I'd like to get rid of this one in favor of what you are proposing, but would like to move the rest of the things though first. And maintain this test meanwhile. WDYT?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java (hunk 5 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java]: IntelliJ complains about fields or local variables having `Optional` as a type, and it makes some sense to me that it does, though not always. I'm in two minds about this a bit tbh. But in this case I prefer using `null`.

Another option would be to replace `isTaskHistoryMaintained()` with something like making `UnitOfWork.getExecutionHistoryStore()` return an empty `Optional` when history shouldn't be maintained, so we could use `Optional.map()` here.

Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java]: Thinking about this a bit more I think I'd actually like to make `getExecutionHistoryStore()` into an `Optional`, but that requires more changes, so I'd do it separately.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java (hunk 4 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java]: The fact that these interfaces can currently be implemented as functions is basically just a coincidence, and not something I intend to keep like that necessarily.

The visitors are not supposed to be implemented by the consumers of this API, so they shouldn't care much.

The only use I see of this annotation is that the compiler would tell me if I accidentally(?) added another method. Which doesn't sound very useful in this situation. So I opted to remove the noise.

**Comment by lptr** [comment]
Rebased to remove the unrelated commits.

**Comment by paplorinc** [comment]
I don't understand, didn't we just say we're not going to rebase?
Now most [links](https://github.com/gradle/gradle/pull/9868/files/377f5eeafab7626e2e4b23f61a043ac23623bacf#diff-b58695f258ac9212686bd1998667290e) are not working.
I wouldn't mind, but it seemed that this was important to you.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy]: Sounds good to me.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java (hunk 5 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java]: Fine with me.
